chain.go

```diff
@@ -0,0 +1,304 @@
+		Variables map[string]interface{} `json:"variables"`
+	}
+
+	ChainOption func(*chain)
+)
+
+func NewChain(endpoint string, opts ...ChainOption) RequestChain {
+	c := &chain{
+		endpoint:   endpoint,
+		httpClient: http.DefaultClient,
+		log:        func(string) {},
+	}
+	for _, optionFunc := range opts {
+		optionFunc(c)
+	}
+	return c
+}
+
+// WithHTTPClient specifies the underlying http.Client to use when
+// making requests.
+//  NewClient(endpoint, WithHTTPClient(specificHTTPClient))
+func WithHTTPClient(httpclient CustomHTTPClient) ChainOption {
+	return func(c *chain) {
+		c.httpClient = httpclient
+	}
+}
+
+// UseMultipartForm uses multipart/form-data and activates support for
+// files.
+func UseMultipartForm() ChainOption {
+	return func(c *chain) {
+		c.useMultipartForm = true
+	}
+}
+
+// ImmediatelyCloseReqBody will close the req body immediately after each request body is ready
+func ImmediatelyCloseReqBody() ChainOption {
+	return func(c *chain) {
+		c.closeReq = true
+	}
+}
+
+func (d *chain) Action(o Operation) (*GraphResponse, Error) {
+	return d.start(o).
+		validate().
+		createRequest().
+		call().
+		parseResponse()
+}
+
+func (d *chain) start(o Operation) *chainState {
+	return &chainState{
+		origin:    d,
+		operation: o,
+	}
+}
+
+func (c *chainState) validate() *chainState {
+	if len(c.operation.Request().Files()) > 0 && !c.origin.useMultipartForm {
+		c.error = NewExecutionError(errors.New("cannot send files with PostFields option"))
+	}
+	return c
+}
+
+func (c *chainState) createRequest() *chainState {
+	if c.error == nil {
+		request := c.operation.Request()
+		requestBody, mimeType, err := c.createRequestBody()
+		if err != nil {
+			c.error = err
+		}
+		r, e := http.NewRequest(http.MethodPost, c.origin.endpoint, requestBody)
+		if e != nil {
+			c.error = NewExecutionError(e)
+		}
+		r.Close = c.origin.closeReq
+		r.Header.Set("Content-Type", mimeType)
+		r.Header.Set("Accept", "application/json; charset=utf-8")
+		for key, values := range request.Headers() {
+			for _, value := range values {
+				r.Header.Add(key, value)
+			}
+		}
+		c.logf(">> headers: %v", r.Header)
+		c.request = r
+	}
+	return c
+}
+
+func (c *chainState) call() *chainState {
+	if c.error == nil {
+		response, err := c.origin.httpClient.Do(c.request)
+		if err == nil {
+			if response.StatusCode != http.StatusOK {
+				c.error = NewHTTPRequestError(response)
+			}
+			c.response = response
+		} else {
+			c.error = NewExecutionResponseError(err, response)
+		}
+	}
+	return c
+}
+
+func (c *chainState) parseResponse() (*GraphResponse, Error) {
+	if c.error == nil {
+		defer c.response.Body.Close()
+		var buf bytes.Buffer
+		if _, err := io.Copy(&buf, c.response.Body); err != nil {
+			return nil, NewExecutionResponseError(errors.Wrap(err, "reading body"), c.response)
+		}
+		c.logf("<< %s", buf.String())
+		resp := c.operation.ResponseBodyAs()
+		var gr *GraphResponse
+		if c.operation.IsMutation() {
```

Comment:
I created the `Mutation` struct to make the `result.Successful` field control if the request was of the `Request` (query) or `Mutation` type and based on the type of the request this flow decided what flow was gonna be used. If you think is easier to control this by a boolean flag we could also remove some of that logic and merge the `Operation` and `Request` interfaces, wdyt?